File: generated_code/GeminiPro/Java/055_SplayTree.java
/**
 * A complete, self-contained Java class that implements an integer set using a Splay Tree.
 * A splay tree is a self-balancing binary search tree with the additional property that
 * recently accessed elements are quick to access again. It performs basic operations
 * such as insertion, look-up and removal in O(log n) amortized time.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     * It is a private static nested class as it is tightly coupled with the SplayTree class.
     */
    private static class Node {
        int key;
        Node parent;
        Node left;
        Node right;

        /**
         * Constructs a new node.
         * @param key The integer key for the node.
         * @param parent The parent of this node.
         */
        Node(int key, Node parent) {
            this.key = key;
            this.parent = parent;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a left rotation on the given node x.
     *   x           y
     *  / \         / \
     * A   y  -->  x   C
     *    / \     / \
     *   B   C   A   B
     * @param x The node to rotate.
     */
    private void leftRotate(Node x) {
        Node y = x.right;
        if (y != null) {
            x.right = y.left;
            if (y.left != null) {
                y.left.parent = x;
            }
            y.parent = x.parent;
            if (x.parent == null) {
                this.root = y;
            } else if (x == x.parent.left) {
                x.parent.left = y;
            } else {
                x.parent.right = y;
            }
            y.left = x;
            x.parent = y;
        }
    }

    /**
     * Performs a right rotation on the given node y.
     *     y         x
     *    / \       / \
     *   x   C --> A   y
     *  / \           / \
     * A   B         B   C
     * @param y The node to rotate.
     */
    private void rightRotate(Node y) {
        Node x = y.left;
        if (x != null) {
            y.left = x.right;
            if (x.right != null) {
                x.right.parent = y;
            }
            x.parent = y.parent;
            if (y.parent == null) {
                this.root = x;
            } else if (y == y.parent.left) {
                y.parent.left = x;
            } else {
                y.parent.right = x;
            }
            x.right = y;
            y.parent = x;
        }
    }

    /**
     * Performs the splay operation on a node x, moving it to the root.
     * @param x The node to splay.
     */
    private void splay(Node x) {
        if (x == null) return;

        while (x.parent != null) {
            Node parent = x.parent;
            Node grandParent = parent.parent;

            if (grandParent == null) { // Zig step
                if (x == parent.left) {
                    rightRotate(parent);
                } else {
                    leftRotate(parent);
                }
            } else { // Zig-Zig or Zig-Zag
                if (x == parent.left) {
                    if (parent == grandParent.left) { // Zig-Zig (left-left)
                        rightRotate(grandParent);
                        rightRotate(parent);
                    } else { // Zig-Zag (right-left)
                        rightRotate(parent);
                        leftRotate(grandParent);
                    }
                } else { // x is right child
                    if (parent == grandParent.left) { // Zig-Zag (left-right)
                        leftRotate(parent);
                        rightRotate(grandParent);
                    } else { // Zig-Zig (right-right)
                        leftRotate(grandParent);
                        leftRotate(parent);
                    }
                }
            }
        }
    }

    /**
     * A helper method to find a node with a given key, or the last node
     * accessed on the search path. It then splays this node to the root.
     *
     * @param key The key to search for.
     * @return The new root of the tree after splaying.
     */
    private Node findAndSplay(int key) {
        if (root == null) {
            return null;
        }
        Node current = root;
        while (true) {
            if (key < current.key) {
                if (current.left == null) break;
                current = current.left;
            } else if (key > current.key) {
                if (current.right == null) break;
                current = current.right;
            } else {
                break; // Found the key
            }
        }
        splay(current);
        return this.root;
    }

    /**
     * Searches for a key in the tree. If the key is found, the node containing it
     * is splayed to the root. If not found, the last accessed node in the search
     * path is splayed to the root.
     *
     * @param key The key to search for.
     * @return The key if it's found, otherwise null.
     */
    public Integer search(int key) {
        Node newRoot = findAndSplay(key);
        if (newRoot != null && newRoot.key == key) {
            return newRoot.key;
        }
        return null;
    }

    /**
     * Inserts a key into the splay tree. If the key already exists, the node
     * containing it is splayed to the root. Otherwise, a new node is inserted
     * and then splayed to the root.
     *
     * @param key The key to insert.
     */
    public void insert(int key) {
        if (root == null) {
            root = new Node(key, null);
            return;
        }

        Node current = root;
        Node parent = null;
        while (current != null) {
            parent = current;
            if (key < current.key) {
                current = current.left;
            } else if (key > current.key) {
                current = current.right;
            } else {
                // Key already exists. Splay it and we're done.
                splay(current);
                return;
            }
        }

        // Key not found, insert it.
        Node newNode = new Node(key, parent);
        if (key < parent.key) {
            parent.left = newNode;
        } else {
            parent.right = newNode;
        }
        splay(newNode);
    }

    /**
     * Deletes a key from the splay tree. The standard splay tree deletion algorithm is used:
     * 1. Splay the node with the key to the root.
     * 2. If the key is not found, do nothing.
     * 3. If found, remove the root, splitting the tree into two subtrees (L and R).
     * 4. Find the maximum element in the left subtree L.
     * 5. Splay this maximum element to the root of L.
     * 6. Attach R as the right child of the new root of L.
     *
     * @param key The key to delete.
     */
    public void delete(int key) {
        findAndSplay(key); // Splay the node (or its neighbor) to the root

        // If root is null or key is not at the root, the key wasn't in the tree.
        if (root == null || root.key != key) {
            return;
        }

        Node leftSubtree = root.left;
        Node rightSubtree = root.right;

        if (leftSubtree == null) {
            root = rightSubtree;
            if (root != null) {
                root.parent = null;
            }
        } else {
            leftSubtree.parent = null;

            // Find the maximum node in the left subtree
            Node maxNode = leftSubtree;
            while (maxNode.right != null) {
                maxNode = maxNode.right;
            }

            // Temporarily set the tree's root to the left subtree to use the splay method
            this.root = leftSubtree;
            splay(maxNode); // maxNode is now the root of this subtree

            // Re-attach the right subtree. The new root (maxNode) has no right child.
            this.root.right = rightSubtree;
            if (rightSubtree != null) {
                rightSubtree.parent = this.root;
            }
        }
    }
}

File: generated_code/GeminiPro/Python/055_SplayTree.py
import sys
from typing import Optional

class SplayTree:
    """
    Implements a Splay Tree, a self-balancing binary search tree.

    This class provides a dictionary-like set for integers with
    insert, delete, and search operations. The key feature is that
    any accessed node is moved to the root of the tree through a series
    of rotations, a process called "splaying". This ensures that
    frequently or recently accessed elements are quick to find.
    """

    class _Node:
        """A node in the Splay Tree."""
        def __init__(self, key: int, parent: Optional['SplayTree._Node'] = None):
            self.key = key
            self.parent = parent
            self.left: Optional['SplayTree._Node'] = None
            self.right: Optional['SplayTree._Node'] = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root: Optional[SplayTree._Node] = None

    def _left_rotate(self, x: _Node):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x: _Node):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, node: _Node):
        """
        Performs the splay operation on a node, moving it to the root.
        """
        while node.parent:
            parent = node.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if node == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.left:  # Zig-Zig case
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif node == parent.right and parent == grandparent.right:  # Zig-Zig case
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif node == parent.left and parent == grandparent.right:  # Zig-Zag case
                self._right_rotate(parent)
                self._left_rotate(grandparent)
            else:  # Zig-Zag case
                self._left_rotate(parent)
                self._right_rotate(grandparent)

    def search(self, key: int) -> bool:
        """
        Searches for a key in the tree.
        
        The accessed node (if found) or its parent (if the key is not found)
        is splayed to become the new root.

        Args:
            key: The integer key to search for.

        Returns:
            True if the key is found, False otherwise.
        """
        if not self.root:
            return False

        current = self.root
        last_node = None
        while current:
            last_node = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key found, splay the node and return
                self._splay(current)
                return True

        # Key not found, splay the last visited node (the would-be parent)
        if last_node:
            self._splay(last_node)
        return False

    def insert(self, key: int):
        """
        Inserts a key into the tree.
        
        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and the new node is then splayed
        to the root.

        Args:
            key: The integer key to insert.
        """
        if not self.root:
            self.root = self._Node(key)
            return

        current = self.root
        parent = None
        while current:
            parent = current
            if key < current.key:
                current = current.left
            elif key > current.key:
                current = current.right
            else:
                # Key already exists, splay it and return
                self._splay(current)
                return

        # Insert the new node
        new_node = self._Node(key, parent)
        if key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key: int):
        """
        Deletes a key from the tree.
        
        This method first searches for the key, which splays the accessed
        node (or its parent) to the root. If the key is found at the root,
        it is removed, and its two subtrees are joined.

        Args:
            key: The integer key to delete.
        """
        if not self.root:
            return

        # search() will splay the node with the key (or its parent) to the root
        self.search(key)

        # If the key is not at the root after splaying, it wasn't in the tree
        if self.root.key != key:
            return

        # The node to delete is now the root.
        # Split the tree into two subtrees: L < key and R > key
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # Promote the right subtree to be the new tree
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        elif not right_subtree:
            # Promote the left subtree to be the new tree
            self.root = left_subtree
            if self.root:
                self.root.parent = None
        else:
            # Join the two subtrees.
            # Detach the left subtree to operate on it
            left_subtree.parent = None
            
            # Find the maximum element in the left subtree.
            max_node = left_subtree
            while max_node.right:
                max_node = max_node.right

            # Splay this maximum node to the root of the (temporary) left subtree.
            self.root = left_subtree
            self._splay(max_node)

            # Attach the original right subtree. After splaying, the new root
            # (max_node) has no right child, so we can attach it there.
            self.root.right = right_subtree
            right_subtree.parent = self.root